Sources/TSCUtility/Version.swift

```diff
@@ -103,47 +118,68 @@ extension Version: CustomStringConvertible {
     }
 }
 
-public extension Version {
-
-    /// Create a version object from string.
-    ///
-    /// - Parameters:
-    ///   - string: The string to parse.
-    init?(string: String) {
-        let prereleaseStartIndex = string.firstIndex(of: "-")
-        let metadataStartIndex = string.firstIndex(of: "+")
-
-        let requiredEndIndex = prereleaseStartIndex ?? metadataStartIndex ?? string.endIndex
-        let requiredCharacters = string.prefix(upTo: requiredEndIndex)
-        let requiredComponents = requiredCharacters
-            .split(separator: ".", maxSplits: 2, omittingEmptySubsequences: false)
-            .map(String.init).compactMap({ Int($0) }).filter({ $0 >= 0 })
-
-        guard requiredComponents.count == 3 else { return nil }
-
-        self.major = requiredComponents[0]
-        self.minor = requiredComponents[1]
-        self.patch = requiredComponents[2]
-
-        func identifiers(start: String.Index?, end: String.Index) -> [String] {
-            guard let start = start else { return [] }
-            let identifiers = string[string.index(after: start)..<end]
-            return identifiers.split(separator: ".").map(String.init)
+extension Version: LosslessStringConvertible {
+    /// Initializes a version struct with the provided version string.
+    /// - Parameter version: A version string to use for creating a new version struct.
+    public init?(_ versionString: String) {
```

Comment:
We should make this throwing instead.

Reply:
I forgot where I read it, but my understanding is that for initializations where the error is clear, the initializer should be failable instead of throwing. I think the error in this case is clear: If initialization fails, it's because the input doesn't follow semvar 2.0.

Additionally, `init?` is a requirement for `LosslessStringConvertible`, and a throwing `init` can't satisfy a non-throwing `init?` requirement.

Reply:
I think it would be preferable to still have an error message, so that it is apparent why is is failing (e.g. because it doesn't follow semver 2.0). Having a throwing initializer also gives us the opportunity to spell out exactly which part is wrong.

For `LosslessStringConvertible`, we could just call the throwing initializer and return `nil` if there was an error. This gives clients the ability to choose the throwing initializer if they don't need the `LosslessStringConvertible` conformance.

Reply:
cc @WowbaggersLiquidLunch  ^^

Reply:
Thanks for the ping. I can add an additional initializer that throws. I just haven't got much time to do it lately. I can probably add it some time this week, though.

Reply:
> For `LosslessStringConvertible`, we could just call the throwing initializer and return `nil` if there was an error. This gives clients the ability to choose the throwing initializer if they don't need the `LosslessStringConvertible` conformance.

Actually I might have read this wrong. @neonichu Do you mean something like this:

```swift
extension Version: LosslessStringConvertible {
    init?(_ description: String) throws {
        ...
        if badInput { return nil }
    }
}
```

This initializer doesn't satisfy `LosslessStringConvertible`'s requirement, because a throwing function can not satisfy a non-throwing requirement.

The only way to both have a throwing initializer and satisfy `LosslessStringConvertible`'s requirement is to have both a throwing and a non-throwing initializers.

Reply:
I meant the latter, having two initializers: the throwing one being the "primary" one and then a fallible one that catches the error and returns `nil` to satisfy `LosslessStringConvertible's`.

Reply:
I just force-pushed 4c4689c27677b57df49345d26c03c573b2a22163 which includes the additional throwing initializer. It's also rebased on the main branch, since it's been a long time since the previous push.